sql/colfetcher: TestScanBatchSize failed

[(sql/colfetcher).TestScanBatchSize failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2811441&tab=buildLog) on [master@3632098d5a324b03eab36849546ca858b60255b8](https://github.com/cockroachdb/cockroach/commits/3632098d5a324b03eab36849546ca858b60255b8):

```
=== RUN   TestScanBatchSize
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize072869390
    test_log_scope.go:74: use -show-logs to present logs inline
    vectorized_batch_size_test.go:67: 
        	Error Trace:	vectorized_batch_size_test.go:67
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 9
        	Test:       	TestScanBatchSize
        	Messages:   	should use just 1 batch to scan 511 rows
    vectorized_batch_size_test.go:74: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize072869390
--- FAIL: TestScanBatchSize (0.47s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestScanBatchSize PKG=./pkg/sql/colfetcher TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestScanBatchSize.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colfetcher).TestScanBatchSize failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2812958&tab=buildLog) on [master@53425a18f2330576724391399eb4a820ad16692f](https://github.com/cockroachdb/cockroach/commits/53425a18f2330576724391399eb4a820ad16692f):

```
=== RUN   TestScanBatchSize
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize015192342
    test_log_scope.go:74: use -show-logs to present logs inline
    vectorized_batch_size_test.go:67: 
        	Error Trace:	vectorized_batch_size_test.go:67
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 9
        	Test:       	TestScanBatchSize
        	Messages:   	should use just 1 batch to scan 511 rows
    vectorized_batch_size_test.go:74: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize015192342
--- FAIL: TestScanBatchSize (0.24s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestScanBatchSize PKG=./pkg/sql/colfetcher TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestScanBatchSize.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colfetcher).TestScanBatchSize failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2812959&tab=buildLog) on [master@53425a18f2330576724391399eb4a820ad16692f](https://github.com/cockroachdb/cockroach/commits/53425a18f2330576724391399eb4a820ad16692f):

```
=== RUN   TestScanBatchSize
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize749821239
    test_log_scope.go:74: use -show-logs to present logs inline
    vectorized_batch_size_test.go:67: 
        	Error Trace:	vectorized_batch_size_test.go:67
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 9
        	Test:       	TestScanBatchSize
        	Messages:   	should use just 1 batch to scan 511 rows
    vectorized_batch_size_test.go:74: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize749821239
--- FAIL: TestScanBatchSize (0.93s)

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=4

```
make stressrace TESTS=TestScanBatchSize PKG=./pkg/sql/colfetcher TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestScanBatchSize.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colfetcher).TestScanBatchSize failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2814394&tab=buildLog) on [master@aaac3aeae88c5864643881c7d74e24074ffb2c86](https://github.com/cockroachdb/cockroach/commits/aaac3aeae88c5864643881c7d74e24074ffb2c86):

```
=== RUN   TestScanBatchSize
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize887759008
    test_log_scope.go:74: use -show-logs to present logs inline
    vectorized_batch_size_test.go:67: 
        	Error Trace:	vectorized_batch_size_test.go:67
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 9
        	Test:       	TestScanBatchSize
        	Messages:   	should use just 1 batch to scan 511 rows
    vectorized_batch_size_test.go:74: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize887759008
--- FAIL: TestScanBatchSize (0.50s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestScanBatchSize PKG=./pkg/sql/colfetcher TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestScanBatchSize.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colfetcher).TestScanBatchSize failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2814762&tab=buildLog) on [master@aaac3aeae88c5864643881c7d74e24074ffb2c86](https://github.com/cockroachdb/cockroach/commits/aaac3aeae88c5864643881c7d74e24074ffb2c86):

```
=== RUN   TestScanBatchSize
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize368259749
    test_log_scope.go:74: use -show-logs to present logs inline
    vectorized_batch_size_test.go:67: 
        	Error Trace:	vectorized_batch_size_test.go:67
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 9
        	Test:       	TestScanBatchSize
        	Messages:   	should use just 1 batch to scan 511 rows
    vectorized_batch_size_test.go:74: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize368259749
--- FAIL: TestScanBatchSize (0.62s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestScanBatchSize PKG=./pkg/sql/colfetcher TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestScanBatchSize.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(sql/colfetcher).TestScanBatchSize failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2815293&tab=buildLog) on [master@77f359dd174f89685a8960b9f3f6192bd5e11260](https://github.com/cockroachdb/cockroach/commits/77f359dd174f89685a8960b9f3f6192bd5e11260):

```
=== RUN   TestScanBatchSize
    test_log_scope.go:73: test logs captured to: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize270765307
    test_log_scope.go:74: use -show-logs to present logs inline
    vectorized_batch_size_test.go:67: 
        	Error Trace:	vectorized_batch_size_test.go:67
        	Error:      	Not equal: 
        	            	expected: 1
        	            	actual  : 9
        	Test:       	TestScanBatchSize
        	Messages:   	should use just 1 batch to scan 511 rows
    vectorized_batch_size_test.go:74: -- test log scope end --
test logs left over in: /go/src/github.com/cockroachdb/cockroach/artifacts/logTestScanBatchSize270765307
--- FAIL: TestScanBatchSize (0.59s)

```

<details><summary>More</summary><p>
Parameters:

- GOFLAGS=-json

```
make stressrace TESTS=TestScanBatchSize PKG=./pkg/sql/colfetcher TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestScanBatchSize.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/sql/colfetcher/BUILD.bazel
+++ b/pkg/sql/colfetcher/BUILD.bazel
@@ -59,6 +59,7 @@ go_test(
         "//pkg/security",
         "//pkg/security/securitytest",
         "//pkg/server",
+        "//pkg/testutils",
         "//pkg/testutils/serverutils",
         "//pkg/testutils/skip",
         "//pkg/testutils/testcluster",

--- a/pkg/sql/colfetcher/vectorized_batch_size_test.go
+++ b/pkg/sql/colfetcher/vectorized_batch_size_test.go
@@ -12,12 +12,14 @@ package colfetcher_test
 
 import (
 	"context"
+	"fmt"
 	"regexp"
 	"strconv"
 	"strings"
 	"testing"
 
 	"github.com/cockroachdb/cockroach/pkg/base"
+	"github.com/cockroachdb/cockroach/pkg/testutils"
 	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/testcluster"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
@@ -45,30 +47,38 @@ func TestScanBatchSize(t *testing.T) {
 
 	conn := tc.Conns[0]
 
-	_, err := conn.ExecContext(ctx, `CREATE TABLE t (a PRIMARY KEY) AS SELECT generate_series(1, 511)`)
+	_, err := conn.ExecContext(ctx, `CREATE TABLE t (a PRIMARY KEY) AS SELECT generate_series(1, 511); ANALYZE t`)
 	assert.NoError(t, err)
 
-	rows, err := conn.QueryContext(ctx, `EXPLAIN ANALYZE (VERBOSE, DISTSQL) SELECT * FROM t`)
-	assert.NoError(t, err)
-	batchCountRegex := regexp.MustCompile(`vectorized batch count: (\d+)`)
-	var found bool
-	var sb strings.Builder
-	for rows.Next() {
-		var res string
-		assert.NoError(t, rows.Scan(&res))
-		sb.WriteString(res)
-		sb.WriteByte('\n')
-		matches := batchCountRegex.FindStringSubmatch(res)
-		if len(matches) == 0 {
-			continue
-		}
-		foundBatches, err := strconv.Atoi(matches[1])
+	testutils.SucceedsSoon(t, func() error {
+		rows, err := conn.QueryContext(ctx, `EXPLAIN ANALYZE (VERBOSE, DISTSQL) SELECT * FROM t`)
 		assert.NoError(t, err)
-		assert.Equal(t, 1, foundBatches, "should use just 1 batch to scan 511 rows")
-		found = true
-		break
-	}
-	if !found {
-		t.Fatalf("expected to find a vectorized batch count; found nothing. text:\n%s", sb.String())
-	}
+		batchCountRegex := regexp.MustCompile(`vectorized batch count: (\d+)`)
+		var found, failed bool
+		var foundBatches int
+		var sb strings.Builder
+		for rows.Next() {
+			var res string
+			assert.NoError(t, rows.Scan(&res))
+			sb.WriteString(res)
+			sb.WriteByte('\n')
+			matches := batchCountRegex.FindStringSubmatch(res)
+			if len(matches) == 0 {
+				continue
+			}
+			foundBatches, err = strconv.Atoi(matches[1])
+			assert.NoError(t, err)
+			if foundBatches != 1 {
+				failed = true
+			}
+			found = true
+		}
+		if failed {
+			return fmt.Errorf("should use just 1 batch to scan 511 rows, found %d:\n%s", foundBatches, sb.String())
+		}
+		if !found {
+			t.Fatalf("expected to find a vectorized batch count; found nothing. text:\n%s", sb.String())
+		}
+		return nil
+	})
 }

